doc/ccanlint.1
'\" t
.\"     Title: ccanlint
.\"    Author: [see the "AUTHOR" section]
.\" Generator: DocBook XSL Stylesheets v1.78.1 <http://docbook.sf.net/>
.\"      Date: 09/28/2015
.\"    Manual: \ \&
.\"    Source: \ \&
.\"  Language: English
.\"
.TH "CCANLINT" "1" "09/28/2015" "\ \&" "\ \&"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
ccanlint \- Make CCAN code modules, and the brightness up\&.
.SH "SYNOPSIS"
.sp
\fBccanlint\fR [\fIOPTIONS\fR] [\fIDIRECTORY\fR\&...]
.SH "DESCRIPTION"
.sp
No encoder? No need to \fBccanlint\fR\&. You programmer? Excited to \fBccanlint\fR!
.sp
CCAN module is small code of the song\&. \fBccanlint\fR full CCAN testing tool\&. Each test spray bit of wisdom\&. Also score\&. Good score good\&. Bad score bad\&.
.sp
\fBccanlint\fR expect the source code in this directory, or command line can be more than one\&. Exit 0 happy if all modules all tests happy\&.
.SH "OPTIONS"
.PP
\fB\-v, \-\-verbose\fR
.RS 4
Make
\fBccanlint\fR
talkative\&. "\-vv" doing very talkative\&. "\-vvvv" make stupid talker\&.
.RE
.PP
\fB\-n, \-\-safe\-mode\fR
.RS 4
Do not compile anything\&. Could it be safer for the bad code, but
\fBccanlint\fR
sad useless\&.
.RE
.PP
\fB\-l, \-\-list\-tests\fR
.RS 4
Tests show
\fBccanlint\fR
can do\&. Then die happy\&.
.RE
.PP
\fB\-\-test\-dep\-graph\fR
.RS 4
Chart of all parties
\fBccanlint\fR
tests
\fIdot(1)\fR
Graphviz, then die happy\&.
.RE
.PP
\fB\-k, \-\-keep\fR
.RS 4
\fBccanlint\fR
normally make mess temporary directory, but now it later in forensic\&.
.RE
.PP
\fB\-s, \-\-summary\fR
.RS 4
\fBccanlint\fR
just realized there is no message unless you die horrible\&.
.RE
.PP
\fB\-x, \-\-exclude\fR=\fITESTNAME\fR
.RS 4
No test run\&. Can the use of time many, many do
\fBccanlint\fR
very, very quickly\&. Often hatred
\fItests_run_valgrind\fR
that the test slowed\&.
.RE
.PP
\fB\-\-timeout\fR=\fIMILLISECONDS\fR
.RS 4
Stop the test and forget it if you take too long\&. Generally, the same works as
\fI\-x tests_run_valgrind\fR\&.
.RE
.PP
\fB\-t, \-\-target\fR=\fITESTNAME\fR
.RS 4
Do not run all tests\&. Run this test, and the proof you need\&. Used many times for many tests\&.
.RE
.PP
\fB\-\-compiler\fR=\fICOMPILER\fR
.RS 4
\fBccanlint\fR
read config\&.h about finding
\fICCAN_COMPILER\fR\&. Otherwise use the default when it was built\&. The change, to use this compiler\&.
.RE
.PP
\fB\-\-cflags\fR=\fICFLAGS\fR
.RS 4
Set compiler options to compile\&. Be sure to protect spaces shell hunger\&.
.RE
.SH "TESTS"
.sp
\fBccanlint\fR many tests\&. Each test will score soon\&. Not total score for each test unless stupid module or no life\&. Worry if little or low score score after hacking\&.
.sp
If test break, but not repair, or maybe the dumb test, put the magic lines \fI_info\fR file like this\&. \fBccanlint\fR to score from 0 of 1 for test, but happy:
.sp
.if n \{\
.RS 4
.\}
.nf
 * Ccanlint:
 *      // Test module foolish for me great!
 *      info_documentation_exists FAIL
 *      // Error for the file may be only
 *      tests_pass_valgrind_noleaks test/run\-mytest\&.c:FAIL
.fi
.if n \{\
.RE
.\}
.PP
\fBinfo_exists\fR
.RS 4
CCAN module must have
\fI_info\fR
file describing\&. No this score is 0\&. However,
\fBccanlint\fR
question may help to write one\&.
.RE
.PP
\fBdepends_exist\fR
.RS 4
\fI_info\fR
file CCAN other module without saying, must find\&. It is not score 0\&.
.RE
.PP
\fBobjects_build\fR
.RS 4
All build purposes
\fI\&.c\fR
in the top dir\&. Not score 0\&.
.RE
.PP
\fBmodule_builds\fR
.RS 4
Link to all objects in an object module\&. Not score 0\&.
.RE
.PP
\fBdepends_accurate\fR
.RS 4
Include other CCAN modules, we must say we need to
\fI_info\fR
depends\&. Only one thing allows different, you can use
\fIccan/tap\fR
for testing anyway\&.
.RE
.PP
\fBdepends_build\fR
.RS 4
We try to generate the CCAN module you need\&.
.RE
.PP
\fBexamples_exist\fR
.RS 4
Rather hope that the comments in the header, and
\fI_info\fR\&. An example of the section in each, please! Maybe more,
\fBccanlint\fR
very happy morning\&.
.RE
.PP
\fBexamples_relevant\fR
.RS 4
Example, do not cut and paste away! You say the name of the thing in the example or
\fBccanlint\fR
unhappy\&.
.RE
.PP
\fBhash_if\fR
.RS 4
Module wants
\fBccanlint\fR\fIconfig\&.h\fR
"#define HAVE_FEATURE" for all feature\&. Function test "#if HAVE_FEATURE" no "#ifdef HAVE_FEATURE" because user might not know about the role at all\&. Intelligent GCC flag
\fI\-Wundef\fR
say HAVE_FEATURE not 0, not 1! but only if the use of
\fI#if\fR\&.
.RE
.PP
\fBinfo_documentation_exists\fR
.RS 4
\fI_info\fR
file format is pretty comments\&. Copying someone\&. It is not difficult write documentation!
.RE
.PP
\fBinfo_summary_single_line\fR
.RS 4
Comments from a top line often describe the function or macro\&.
\fI_info\fR
comment top line describes complete module\&. Characteristics make you scream!
.RE
.PP
\fBlicense_exists\fR
.RS 4
The lawyers eat me\&.
\fI_info\fR
have
\fILicense:\fR
in the observation and LICENSE file there\&. In general, is the link:
\fBccanlint\fR
offer create a link, if they know
\fILicense:\fR\&.
.RE
.PP
\fBlicense_comment\fR
.RS 4
Attorney everywhere\&. Please put a comment saying something like "GPL Version 4\&. Read LICENSE\&." in all source files in the directory\&.
.RE
.PP
\fBlicense_file_compat\fR
.RS 4
Do not lie about the license!
\fBccanlint\fR
search files, see the license of another, angry here\&.
.RE
.PP
\fBlicense_depends_compat\fR
.RS 4
Hostile to BSD license module, but requires another module of the GPL\&. Perhaps poor encoder think all BSD code, unloading and damage attorney stick! Ay ay!
.RE
.PP
\fBmain_header_exists\fR
.RS 4
\fBccanlint\fR
know the module name directory name\&. Expect the same name for header\&.
.RE
.PP
\fBheaders_idempotent\fR
.RS 4
Good header
\fI#include\fR
many time happy\&. Rap header around easy\&.
\fBccanlint\fR
say it can fix too\&. Always work\&.
.RE
.PP
\fBmain_header_compiles\fR
.RS 4
Simple program
\fI#include\fR
main header compile\&.
.RE
.PP
\fBavoids_cpp_reserved\fR
.RS 4
C++ programmer to include code\&. Not like them anyway, maybe, but wrong end your program, do mourn\&. Only main header compile C++ and if trying to compile C\e++ module stupid to pieces\&.
.RE
.PP
\fBno_trailing_whitespace\fR
.RS 4
Linux kernel programmers more, solve the problem for the space of the final ban the line\&. Now all lots of hackers working to fix it\&. Want to famous and Linux? Leave extra space too!
.RE
.PP
\fBexamples_compile\fR
.RS 4
\fBccanlint\fR
very smart! Take
\fIExample:\fR
from a comment in the header and
\fI_info\fR\&. First try to compile anything\&. If not, add many headers and maybe put inside the function\&. It does not work, adds the latest example\&. If the last example has
\fI\&...\fR
try that maybe\&. Sometimes too complicated!
\fI\-vv\fR
or
\fI\-\-keep\fR
to see why it broke\&. Or maybe bad example
\fBccanlint\fR
says wow!
.RE
.PP
\fBexamples_run\fR
.RS 4
If the example program that comments like
\fI// Given "foo" outputs "bar"\fR\fBccanlint\fR
will run the program with
\fIfoo\fR
in the command line and standard input\&. Happy if
\fIbar\fR
are out\&. If quotes around
\fIbar\fR
exact match needed; without quotes whitespace matches any other space and trailing ignored\&. \en is also supported for matching\&. You can also
\fI"output contains"\fR
to pass if the output contains the string\&.
.RE
.PP
\fBmodule_links\fR
.RS 4
CCAN link to the program module simply no error\&.
.RE
.PP
\fBobjects_build_with_stringchecks\fR
.RS 4
Module
\fIccan/str\fR
is super difficult to detect errors debugging chain\&.
\fBccanlint\fR
use with the module and see break!
.RE
.PP
\fBtests_exist\fR
.RS 4
You have CCAN module directory called
\fItest\fR\&. You have proof here\&. If there is no proof,
\fBccanlint\fR
still offer make proof for you\&.
.RE
.PP
\fBtests_compile\fR
.RS 4
In
\fItest\fR
which has four such tests, start with different name\&.
\fIrun\fR
compile the test files, but no link to the module, you
\fI#include\fR
to get the bits of the module\&.
\fIapi\fR
test compile and link with the module\&.
\fIcompile\-ok\fR
as
\fIrun\fR
but only build\&.
\fIcompile\-fail\fR
compile, but when
\fIFAIL\fR
set has to break or alert\&. This good for module supposed to warn\&.
.RE
.PP
\fBtest_helpers_compile\fR
.RS 4
Other files
\fItest\fR? Compilation of links to all tests\&. Ask for help\&.
.RE
.PP
\fBtests_pass\fR
.RS 4
\fIrun\fR
and
\fIapi\fR
test happy departure\&. If not happy, offer debugger\&.
.RE
.PP
\fBtests_pass_valgrind\fR
.RS 4
\fBvalgrind\fR
the tool of all
\fIrun\fR
and
\fIapi\fR
slow test\&. However, we found many errors! If
\fBvalgrind\fR
test rest,
\fI_info\fR
have
\fBccanlint\fR
section, make "tests_pass_valgrind test/TESTNAME:FAIL"\&. If required valgrind additional option, "tests_pass_valgrind test/TESTNAME:\-\-option"\&.
.RE
.PP
\fBtests_pass_valgrind_noleaks\fR
.RS 4
\fBvalgrind\fR
complain if the memory leak test\&.
\fI_info\fR
can also be disabled\&.
.RE
.PP
\fBtests_compile_coverage\fR
.RS 4
Compile
\fIrun\fR,
\fIapi\fR
test coverage\&. Fun if not here!
.RE
.PP
\fBtests_coverage\fR
.RS 4
Run tests again, find lines that never try! Half of the lines 1 point 3/4 getting 2 points\&. Limit of 5 points, but the extra point for all lines of evidence\&. Not win unless the module silly or use gimmick
\fIccan/failtest\fR\&.
.RE
.PP
\fBreduce_features\fR
.RS 4
Code use
\fIHAVE_FEATURE\fR
make special config\&.h turned off\&. Not stupid like HAVE_BIG_ENDIAN though!
.RE
.PP
\fBdepends_build_without_features\fR
.RS 4
Make modules CCAN need\&. config\&.h but not more features\&.
.RE
.PP
\fBobjects_build_without_features\fR
.RS 4
Make the module again, but not more features\&.
.RE
.PP
\fBtests_helpers_compile_without_features\fR
.RS 4
Helpers do try again, but not more features\&.
.RE
.PP
\fBtests_compile_without_features\fR
.RS 4
Collect the tests again, but not more features\&.
.RE
.sp
\fBtests_pass_without_features\fR: Run tests again, but not more features\&.
.SH "BUGS"
.sp
\fBccanlint\fR rapid change\&. The bad man, bad page\&.
.SH "AUTHOR"
.sp
Rusty Russell wrote \fBccanlint\fR\&. Helping others, but most break Rusty\&.
.SH "RESOURCES"
.sp
Main web site: https://github\&.com/rustyrussell/ccan
.sp
Wiki: https://github\&.com/rustyrussell/ccan/wiki/
.SH "COPYING"
.sp
This program is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation; either version 2 of the License, or (at your option) any later version\&.


tools/ccanlint/ccanlint.c
/*
 * ccanlint: assorted checks and advice for a ccan package
 * Copyright (C) 2008 Rusty Russell, Idris Soule
 * Copyright (C) 2010 Rusty Russell, Idris Soule
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the Free
 * Software Foundation; either version 2 of the License, or (at your option)
 * any later version.
 *
 *   This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along with
 * this program; if not, write to the Free Software Foundation, Inc., 51
 * Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */
#include "ccanlint.h"
#include "../tools.h"
#include "../read_config_header.h"
#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <err.h>
#include <ctype.h>
#include <ccan/str/str.h>
#include <ccan/take/take.h>
#include <ccan/opt/opt.h>
#include <ccan/foreach/foreach.h>
#include <ccan/cast/cast.h>
#include <ccan/tlist/tlist.h>
#include <ccan/tal/path/path.h>
#include <ccan/strmap/strmap.h>

typedef STRMAP(struct ccanlint *) ccanlint_map_t;

int verbose = 0;
static ccanlint_map_t tests;
bool safe_mode = false;
bool keep_results = false;
bool non_ccan_deps = false;
bool build_failed = false;
static bool targeting = false;
static unsigned int timeout;

const char *config_header;

const char *ccan_dir;

#if 0
static void indent_print(const char *string)
{
	while (*string) {
		unsigned int line = strcspn(string, "\n");
		printf("\t%.*s", line, string);
		if (string[line] == '\n') {
			printf("\n");
			line++;
		}
		string += line;
	}
}
#endif

bool ask(const char *question)
{
	char reply[80];

	printf("%s ", question);
	fflush(stdout);

	return fgets(reply, sizeof(reply), stdin) != NULL
		&& toupper(reply[0]) == 'Y';
}

/* Skip, but don't remove. */
static bool skip_test(struct dgraph_node *node, const char *why)
{
	struct ccanlint *c = container_of(node, struct ccanlint, node);
	c->skip = why;
	return true;
}

static const char *dep_failed(struct manifest *m UNNEEDED)
{
	return "dependency couldn't run";
}

static bool cannot_run(struct dgraph_node *node, void *all UNNEEDED)
{
	struct ccanlint *c = container_of(node, struct ccanlint, node);
	c->can_run = dep_failed;

	return true;
}

struct run_info {
	bool noninteractive;
	unsigned int score, total;
	struct manifest *m;
	const char *prefix;
	bool pass;
};

static bool run_test(struct dgraph_node *n, struct run_info *run)
{
	struct ccanlint *i = container_of(n, struct ccanlint, node);
	unsigned int timeleft;
	struct score *score;

	if (i->done)
		return true;

	score = tal(run->m, struct score);
	list_head_init(&score->per_file_errors);
	score->error = NULL;
	score->pass = false;
	score->score = 0;
	score->total = 1;

	/* We can see skipped things in two cases:
	 * (1) _info excluded them (presumably because they fail).
	 * (2) A prerequisite failed.
	 */
	if (i->skip) {
		if (verbose)
			printf("%s%s: skipped (%s)\n",
			       run->prefix, i->name, i->skip);
		/* Pass us up to the test which failed, not us. */
		score->pass = true;
		goto out;
	}

	if (i->can_run) {
		i->skip = i->can_run(run->m);
		if (i->skip) {
			/* Test doesn't apply, or can't run?  That's OK. */
			if (verbose > 1)
				printf("%s%s: skipped (%s)\n",
				       run->prefix, i->name, i->skip);
			/* Mark our dependencies to skip. */
			dgraph_traverse_from(&i->node, cannot_run, NULL);
			score->pass = true;
			score->total = 0;
			goto out;
		}
	}

	timeleft = timeout ? timeout : default_timeout_ms;
	i->check(run->m, &timeleft, score);
	if (timeout && timeleft == 0) {
		i->skip = "timeout";
		if (verbose)
			printf("%s%s: skipped (%s)\n",
			       run->prefix, i->name, i->skip);
		/* Mark our dependencies to skip. */
		dgraph_traverse_from(&i->node, skip_test,
				     "dependency timed out");
		score->pass = true;
		score->total = 0;
		goto out;
	}

	assert(score->score <= score->total);
	if (!score->pass
	    || (score->score < score->total && verbose)
	    || verbose > 1) {
		printf("%s%s (%s): %s",
		       run->prefix, i->name, i->key,
		       score->pass ? "PASS" : "FAIL");
		if (score->total > 1)
			printf(" (+%u/%u)", score->score, score->total);
		printf("\n");
	}

	if (!score->pass || verbose) {
		if (score->error) {
			printf("%s%s", score->error,
			       strends(score->error, "\n") ? "" : "\n");
		}
	}
	if (!run->noninteractive && score->score < score->total && i->handle)
		i->handle(run->m, score);

	if (!score->pass) {
		/* Skip any tests which depend on this one. */
		dgraph_traverse_from(&i->node, skip_test, "dependency failed");
	}

out:
	run->score += score->score;
	run->total += score->total;

	/* FIXME: Free score. */
	run->pass &= score->pass;
	i->done = true;

	if (!score->pass && i->compulsory) {
		warnx("%s%s failed", run->prefix, i->name);
		run->score = 0;
		return false;
	}
	return true;
}

static void register_test(struct ccanlint *test)
{
	if (!strmap_add(&tests, test->key, test))
		err(1, "Adding test %s", test->key);
	test->options = tal_arr(NULL, char *, 1);
	test->options[0] = NULL;
	dgraph_init_node(&test->node);
}

static bool get_test(const char *member UNNEEDED, struct ccanlint *i,
		     struct ccanlint **ret)
{
	if (tlist_empty(&i->node.edge[DGRAPH_TO])) {
		*ret = i;
		return false;
	}
	return true;
}

/**
 * get_next_test - retrieves the next test to be processed
 **/
static inline struct ccanlint *get_next_test(void)
{
	struct ccanlint *i = NULL;

	strmap_iterate(&tests, get_test, &i);
	if (i)
		return i;

	if (strmap_empty(&tests))
		return NULL;

	errx(1, "Can't make process; test dependency cycle");
}

static struct ccanlint *find_test(const char *key)
{
	return strmap_get(&tests, key);
}

bool is_excluded(const char *name)
{
	return find_test(name)->skip != NULL;
}

static bool init_deps(const char *member UNNEEDED,
		      struct ccanlint *c, void *unused UNNEEDED)
{
	char **deps = tal_strsplit(NULL, c->needs, " ", STR_EMPTY_OK);
	unsigned int i;

	for (i = 0; deps[i]; i++) {
		struct ccanlint *dep;

		dep = find_test(deps[i]);
		if (!dep)
			errx(1, "BUG: unknown dep '%s' for %s",
			     deps[i], c->key);
		dgraph_add_edge(&dep->node, &c->node);
	}
	tal_free(deps);
	return true;
}

static bool check_names(const char *member UNNEEDED, struct ccanlint *c,
			ccanlint_map_t *names)
{
	if (!strmap_add(names, c->name, c))
		err(1, "Duplicate name %s", c->name);
	return true;
}

static void init_tests(void)
{
	ccanlint_map_t names;
	struct ccanlint **table;
	size_t i, num;

	strmap_init(&tests);

	table = autodata_get(ccanlint_tests, &num);
	for (i = 0; i < num; i++)
		register_test(table[i]);
	autodata_free(table);

	strmap_iterate(&tests, init_deps, NULL);

	/* Check for duplicate names. */
	strmap_init(&names);
	strmap_iterate(&tests, check_names, &names);
	strmap_clear(&names);
}

static bool reset_test(struct dgraph_node *node, void *unused UNNEEDED)
{
	struct ccanlint *c = container_of(node, struct ccanlint, node);
	c->skip = NULL;
	c->done = false;
	return true;
}

static void reset_tests(struct dgraph_node *all)
{
	dgraph_traverse_to(all, reset_test, NULL);
}

static bool print_deps(const char *member UNNEEDED,
		       struct ccanlint *c, void *unused UNNEEDED)
{
	if (!tlist_empty(&c->node.edge[DGRAPH_FROM])) {
		struct dgraph_edge *e;

		printf("These depend on %s:\n", c->key);
		dgraph_for_each_edge(&c->node, e, DGRAPH_FROM) {
			struct ccanlint *to = container_of(e->n[DGRAPH_TO],
							   struct ccanlint,
							   node);
			printf("\t%s\n", to->key);
		}
	}
	return true;
}

static void print_test_depends(void)
{
	printf("Tests:\n");

	strmap_iterate(&tests, print_deps, NULL);
}


static void show_tmpdir(const char *dir)
{
	printf("You can find ccanlint working files in '%s'\n", dir);
}

static char *keep_tests(void *unused UNNEEDED)
{
	keep_results = true;

	/* Don't automatically destroy temporary dir. */
	keep_temp_dir();
	tal_add_destructor(temp_dir(), show_tmpdir);
	return NULL;
}

static bool remove_test(struct dgraph_node *node, const char *why)
{
	struct ccanlint *c = container_of(node, struct ccanlint, node);
	c->skip = why;
	dgraph_clear_node(node);
	return true;
}

static char *exclude_test(const char *testname, void *unused UNNEEDED)
{
	struct ccanlint *i = find_test(testname);
	if (!i)
		return tal_fmt(NULL, "No test %s to --exclude", testname);

	/* Remove this, and everything which depends on it. */
	dgraph_traverse_from(&i->node, remove_test, "excluded on command line");
	remove_test(&i->node, "excluded on command line");
	return NULL;
}

static void skip_test_and_deps(struct ccanlint *c, const char *why)
{
	/* Skip this, and everything which depends on us. */
	dgraph_traverse_from(&c->node, skip_test, why);
	skip_test(&c->node, why);
}

static char *list_tests(void *arg UNNEEDED)
{
	struct ccanlint *i;

	printf("Tests:\n");
	/* This makes them print in topological order. */
	while ((i = get_next_test()) != NULL) {
		printf("   %-25s %s\n", i->key, i->name);
		dgraph_clear_node(&i->node);
		strmap_del(&tests, i->key, NULL);
	}
	exit(0);
}

static bool draw_test(const char *member UNNEEDED,
		      struct ccanlint *c, const char *style)
{
	/*
	 * todo: escape labels in case ccanlint test keys have
	 *       characters interpreted as GraphViz syntax.
	 */
	printf("\t\"%p\" [label=\"%s\"%s]\n", c, c->key, style);
	return true;
}

static void test_dgraph_vertices(const char *style)
{
	strmap_iterate(&tests, draw_test, style);
}

static bool draw_edges(const char *member UNNEEDED,
		       struct ccanlint *c, void *unused UNNEEDED)
{
	struct dgraph_edge *e;

	dgraph_for_each_edge(&c->node, e, DGRAPH_FROM) {
		struct ccanlint *to = container_of(e->n[DGRAPH_TO],
						   struct ccanlint,
						   node);
		printf("\t\"%p\" -> \"%p\"\n", c->name, to->name);
	}
	return true;
}

static void test_dgraph_edges(void)
{
	strmap_iterate(&tests, draw_edges, NULL);
}

static char *test_dependency_graph(void *arg UNNEEDED)
{
	puts("digraph G {");

	test_dgraph_vertices("");
	test_dgraph_edges();

	puts("}");

	exit(0);
}

static void add_options(struct ccanlint *test, char **options,
			unsigned int num_options)
{
	unsigned int num;

	if (!test->options)
		num = 0;
	else
		/* -1, because last one is NULL. */
		num = tal_count(test->options) - 1;

	tal_resize(&test->options, num + num_options + 1);
	memcpy(&test->options[num], options, (num_options + 1)*sizeof(char *));
}

void add_info_options(struct manifest *m)
{
	unsigned int i;
	char **info_options = get_ccanlint(m, m->dir, get_or_compile_info);

	for (i = 0; info_options[i]; i++) {
		char **words = tal_strsplit(m, info_options[i], " \t",
					    STR_NO_EMPTY);
		struct ccanlint *test;

		if (!words[0])
			continue;

		test = find_test(words[0]);
		if (!test) {
			warnx("%s: unknown ccanlint test '%s'",
			      m->info_file->fullname, words[0]);
			continue;
		}

		if (!words[1]) {
			warnx("%s: no argument to test '%s'",
			      m->info_file->fullname, words[0]);
			continue;
		}

		/* Known failure? */
		if (strcasecmp(words[1], "FAIL") == 0) {
			if (!targeting)
				skip_test_and_deps(test,
						   "excluded in _info"
						   " file");
		} else {
			if (!test->takes_options)
				warnx("%s: %s doesn't take options",
				      m->info_file->fullname, words[0]);
			add_options(test, words+1, tal_count(words)-1);
		}
	}
}

/* If options are of form "filename:<option>" they only apply to that file */
char **per_file_options(const struct ccanlint *test, struct ccan_file *f)
{
	char **ret;
	unsigned int i, j = 0;

	/* Fast path. */
	if (!test->options[0])
		return test->options;

	ret = tal_arr(f, char *, tal_count(test->options));
	for (i = 0; test->options[i]; i++) {
		char *optname;

		if (!test->options[i] || !strchr(test->options[i], ':')) {
			optname = test->options[i];
		} else if (strstarts(test->options[i], f->name)
			   && test->options[i][strlen(f->name)] == ':') {
			optname = test->options[i] + strlen(f->name) + 1;
		} else
			continue;

		/* FAIL overrides anything else. */
		if (streq(optname, "FAIL")) {
			ret = tal_arr(f, char *, 2);
			ret[0] = (char *)"FAIL";
			ret[1] = NULL;
			return ret;
		}
		ret[j++] = optname;
	}
	ret[j] = NULL;

	/* Shrink it to size so tal_array_length() works as expected. */
	tal_resize(&ret, j + 1);
	return ret;
}

static char *opt_set_const_charp(const char *arg, const char **p)
{
	return opt_set_charp(arg, cast_const2(char **, p));
}

static char *opt_set_target(const char *arg, struct dgraph_node *all)
{
	struct ccanlint *t = find_test(arg);
	if (!t)
		return tal_fmt(NULL, "unknown --target %s", arg);

	targeting = true;
	dgraph_add_edge(&t->node, all);
	return NULL;
}

static bool run_tests(struct dgraph_node *all,
		      bool summary,
		      bool deps_fail_ignore,
		      struct manifest *m,
		      const char *prefix)
{
	struct run_info run;
	const char *comment = "";

	run.noninteractive = summary;
	run.m = m;
	run.prefix = prefix;
	run.score = run.total = 0;
	run.pass = true;

	non_ccan_deps = build_failed = false;

	dgraph_traverse_to(all, run_test, &run);

	/* We can completely fail if we're missing external stuff: ignore */
	if (deps_fail_ignore && non_ccan_deps && build_failed) {
		comment = " (missing non-ccan dependencies?)";
		run.pass = true;
	} else if (!run.pass) {
		comment = " FAIL!";
	}
	printf("%sTotal score: %u/%u%s\n",
	       prefix, run.score, run.total, comment);

	return run.pass;
}

static bool add_to_all(const char *member UNNEEDED, struct ccanlint *c,
		       struct dgraph_node *all)
{
	/* If we're excluded on cmdline, don't add. */
	if (!c->skip)
		dgraph_add_edge(&c->node, all);
	return true;
}

static bool test_module(struct dgraph_node *all,
			const char *dir, const char *prefix, bool summary,
			bool deps_fail_ignore)
{
	struct manifest *m = get_manifest(autofree(), dir);
	char *testlink = path_join(NULL, temp_dir(), "test");

	/* Create a symlink from temp dir back to src dir's
	 * test directory. */
	unlink(testlink);
	if (symlink(path_join(m, dir, "test"), testlink) != 0)
		err(1, "Creating test symlink in %s", temp_dir());

	return run_tests(all, summary, deps_fail_ignore, m, prefix);
}

int main(int argc, char *argv[])
{
	bool summary = false, pass = true, deps_fail_ignore = false;
	int i;
	const char *prefix = "";
	char *cwd = path_cwd(NULL), *dir;
	struct ccanlint top;  /* cannot_run may try to set ->can_run */
	const char *override_compiler = NULL, *override_cflags = NULL;
	const char *override_gcov = NULL;
	
	/* Empty graph node to which we attach everything else. */
	dgraph_init_node(&top.node);

	opt_register_early_noarg("--verbose|-v", opt_inc_intval, &verbose,
				 "verbose mode (up to -vvvv)");
	opt_register_noarg("-n|--safe-mode", opt_set_bool, &safe_mode,
			 "do not compile anything");
	opt_register_noarg("-l|--list-tests", list_tests, NULL,
			 "list tests ccanlint performs (and exit)");
	opt_register_noarg("--test-dep-graph", test_dependency_graph, NULL,
			 "print dependency graph of tests in Graphviz .dot format");
	opt_register_noarg("-k|--keep", keep_tests, NULL,
			 "do not delete ccanlint working files");
	opt_register_noarg("--summary|-s", opt_set_bool, &summary,
			   "give results only, no interactive correction");
	opt_register_arg("-x|--exclude <testname>", exclude_test, NULL, NULL,
			 "exclude <testname> (can be used multiple times)");
	opt_register_arg("--timeout <milleseconds>", opt_set_uintval,
			 NULL, &timeout,
			 "ignore (terminate) tests that are slower than this");
	opt_register_arg("-t|--target <testname>", opt_set_target, NULL,
			 &top.node,
			 "only run one test (and its prerequisites)");
	opt_register_arg("--compiler <compiler>", opt_set_const_charp,
			 NULL, &override_compiler, "set the compiler");
	opt_register_arg("--cflags <flags>", opt_set_const_charp,
			 NULL, &override_cflags, "set the compiler flags");
	opt_register_arg("--gcov <coverage tool>", opt_set_const_charp,
			 NULL, &override_gcov, "set the coverage tool");
	opt_register_noarg("--deps-fail-ignore", opt_set_bool,
			   &deps_fail_ignore,
			   "don't fail if external dependencies are missing");
	opt_register_noarg("-?|-h|--help", opt_usage_and_exit,
			   "\nA program for checking and guiding development"
			   " of CCAN modules.",
			   "This usage message");

	/* Do verbose before anything else... */
	opt_early_parse(argc, argv, opt_log_stderr_exit);

	/* We move into temporary directory, so gcov dumps its files there. */
	if (chdir(temp_dir()) != 0)
		err(1, "Error changing to %s temporary dir", temp_dir());

	init_tests();

	if (verbose >= 3) {
		compile_verbose = true;
		print_test_depends();
	}
	if (verbose >= 4)
		tools_verbose = true;

	opt_parse(&argc, argv, opt_log_stderr_exit);

	if (!targeting)
		strmap_iterate(&tests, add_to_all, &top.node);

	if (argc == 1)
		dir = cwd;
	else
		dir = path_simplify(NULL, take(path_join(NULL, cwd, argv[1])));

	ccan_dir = find_ccan_dir(dir);
	if (!ccan_dir)
		errx(1, "Cannot find ccan/ base directory in %s", dir);
	config_header = read_config_header(ccan_dir, verbose > 1);

	/* We do this after read_config_header has set compiler & cflags */
	if (override_cflags)
		cflags = override_cflags;
	if (override_compiler)
		compiler = override_compiler;
	if (override_gcov)
		gcov = override_gcov;

	if (argc == 1)
		pass = test_module(&top.node, cwd, "",
				   summary, deps_fail_ignore);
	else {
		for (i = 1; i < argc; i++) {
			dir = path_canon(NULL,
					 take(path_join(NULL, cwd, argv[i])));
			if (!dir)
				err(1, "Cannot get canonical name of '%s'",
				    argv[i]);

			prefix = path_join(NULL, ccan_dir, "ccan");
			prefix = path_rel(NULL, take(prefix), dir);
			prefix = tal_strcat(NULL, take(prefix), ": ");

			pass &= test_module(&top.node, dir, prefix, summary,
					    deps_fail_ignore);
			reset_tests(&top.node);
		}
	}
	return pass ? 0 : 1;
}
